internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -2716,28 +2693,36 @@ func validate(ctx context.Context, log logrus.FieldLogger, asc ASC) (bool, error
 		return false, nil
 	}
 
-	// validate kubernetes ingress config if not running on OpenShift
-	// Ingress must not be nil and both hostnames must be provided
-	if !asc.rec.IsOpenShift &&
-		(asc.spec.Ingress == nil ||
-			asc.spec.Ingress.AssistedServiceHostname == "" ||
-			asc.spec.Ingress.ImageServiceHostname == "") {
-		message := "ingress configuration is required for non-OpenShift deployment"
-		log.Error(message)
-		conditionsv1.SetStatusConditionNoHeartbeat(
-			asc.conditions, conditionsv1.Condition{
-				Type:    aiv1beta1.ConditionReconcileCompleted,
-				Status:  corev1.ConditionFalse,
-				Reason:  aiv1beta1.ReasonKubernetesIngressMissing,
-				Message: message,
-			},
-		)
-		err = asc.Client.Status().Update(ctx, asc.Object)
-		if err != nil {
-			log.WithError(err).Error("Failed to update status")
-			return false, err
+	if !asc.rec.IsOpenShift {
+		message := ""
+		reason := ""
+		// validate kubernetes ingress config if not running on OpenShift
+		// Ingress must not be nil and both hostnames must be provided
+		if asc.spec.Ingress == nil || asc.spec.Ingress.AssistedServiceHostname == "" || asc.spec.Ingress.ImageServiceHostname == "" {
+			message = "ingress configuration is required for non-OpenShift deployment"
+			reason = aiv1beta1.ReasonKubernetesIngressMissing
+		} else if !supportsCertManager {
+			message = "cert-manager is a required dependency for non-OpenShift deployments"
+			reason = aiv1beta1.ReasonCertificateFailure
+		}
+
+		if message != "" && reason != "" {
+			log.Error(message)
+			conditionsv1.SetStatusConditionNoHeartbeat(
+				asc.conditions, conditionsv1.Condition{
+					Type:    aiv1beta1.ConditionReconcileCompleted,
+					Status:  corev1.ConditionFalse,
+					Reason:  reason,
+					Message: message,
+				},
+			)
+			err = asc.Client.Status().Update(ctx, asc.Object)
+			if err != nil {
+				log.WithError(err).Error("Failed to update status")
+				return false, err
+			}
+			return false, nil
```

Comment:
If we extract this error handling we can return earlier and avoid else condition, i.e.
```golang
if asc.spec.Ingress == nil || asc.spec.Ingress.AssistedServiceHostname == "" || asc.spec.Ingress.ImageServiceHostname == "" {
		return handleError("ingress configuration is required for non-OpenShift deployment", aiv1beta1.ReasonKubernetesIngressMissing)
}
if !supportsCertManager {
		return handleError("cert-manager is a required dependency for non-OpenShift deployments", aiv1beta1.ReasonCertificateFailure)
}
```

Reply:
Again not sure I see the upside in extracting to a function like `handleError` that only handles errors for this one particular part of validation.

I could be convinced to extract this entire block into something like `validateForKube` if you think this `validate` function is just becoming too complex.

Reply:
This function is almost 100 lines and the "handle error block" is duplicated in lines 2679-2693
Extracting this function would make this validate function way more readable IMO.
It's not too complex, but we would make it easier to follow by exiting earlier and factoring out common verbose code.